snmp/datadog_checks/snmp/models.py

```diff
@@ -16,35 +16,53 @@ class OID(object):
     """
     An SNMP object identifier.
 
-    Acts as a facade for various types used by PySNMP to represent OIDs.
+    Acts as a lazy facade for various types used by PySNMP to represent OIDs.
     """
 
     def __init__(self, value):
         # type: (Union[Sequence[int], str, ObjectName, ObjectIdentity, ObjectType]) -> None
+
+        # NOTE: we can't decode `value` just yet, because it may be a PySNMP object that hasn't been resolved yet.
+        # Passing such unresolved objects here is actually OK: looking up the MIB name of an OID isn't required to query
+        # it from a server. (Eg. we can query '1.3.6.1.2.1.1.2' without saying 'and by the way, this is sysObjectId'.)
+        # Anyway, this is why we do lazy decoding, and why '._initialize()' even exists.
+        self._raw = value
+
+    def _initialize(self):
+        # type: () -> None
+        value = self._raw
+
         try:
             parts = parse_as_oid_tuple(value)
         except CouldNotDecodeOID:
             raise  # Explicitly re-raise this exception.
 
-        # Let's make extra sure we didn't mess up.
-        if not isinstance(parts, tuple):
-            raise RuntimeError(
-                'Expected result {!r} of parsing value {!r} to be a tuple, but got {}'.format(parts, value, type(parts))
-            )  # pragma: no cover
```

Comment:
Dropping this check as we should trust `parse_as_oid_tuple()` — it already fails with `CouldNotDecodeOID` in case of invalid input.